File: rust/day4/src/bin/part2.rs
fn check_mas_from_center(
    grid: &Vec<Vec<char>>,
    center_row: isize,
    center_col: isize,
    dr: isize,
    dc: isize,
) -> bool {
    let rows = grid.len() as isize;
    let cols = grid[0].len() as isize;

    // Check if all positions are within bounds
    for i in -1..=1 {
        let nr = center_row + i * dr;
        let nc = center_col + i * dc;
        if nr < 0 || nr >= rows || nc < 0 || nc >= cols {
            return false;
        }
    }

    // Get the three characters in order
    let chars: Vec<char> = vec![
        grid[(center_row - dr) as usize][(center_col - dc) as usize],
        grid[center_row as usize][center_col as usize],
        grid[(center_row + dr) as usize][(center_col + dc) as usize],
    ];

    // Must be exactly MAS or SAM
    chars == vec!['M', 'A', 'S'] || chars == vec!['S', 'A', 'M']
}

fn count_x_mas(grid: &Vec<Vec<char>>) -> usize {
    let rows = grid.len() as isize;
    let cols = grid[0].len() as isize;
    let mut count = 0;

    // For each possible center point of the X
    for row in 1..rows - 1 {
        for col in 1..cols - 1 {
            // Check if it's an 'A' (center of X)
            if grid[row as usize][col as usize] != 'A' {
                continue;
            }

            // Check both possible X configurations
            let directions = [
                ((1, 1), (1, -1)), // Check both diagonal directions from center
            ];

            'outer: for &(dir1, dir2) in &directions {
                // For each direction pair, try both orientations of MAS/SAM
                let checks = [
                    // Check dir1 as MAS and dir2 as MAS
                    (dir1, dir2),
                    // Check dir1 as SAM and dir2 as SAM
                    ((-dir1.0, -dir1.1), (-dir2.0, -dir2.1)),
                ];

                for &(d1, d2) in &checks {
                    if check_mas_from_center(grid, row as isize, col as isize, d1.0, d1.1)
                        && check_mas_from_center(grid, row as isize, col as isize, d2.0, d2.1)
                    {
                        count += 1;
                        continue 'outer; // Only count each X once
                    }
                }
            }
        }
    }

    count
}

fn read_input(input: &str) -> Vec<Vec<char>> {
    input.lines().map(|line| line.chars().collect()).collect()
}

fn part2(input: &str) -> usize {
    let grid = read_input(input);
    count_x_mas(&grid)
}

fn main() {
    let input = include_str!("../../input.txt");
    let result = part2(input);
    println!("{}", result);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_part2() {
        let input = "MMMSXXMASM
MSAMXMSMSA
AMXSXMAAMM
MSAMASMSMX
XMASAMXAMM
XXAMMXXAMA
SMSMSASXSS
SAXAMASAAA
MAMMMXMMMM
MXMXAXMASX";
        assert_eq!(part2(input), 9);
    }
}

File: ruby/day4/part2.rb
# frozen_string_literal: true

DIAGONAL_DIRECTIONS = [
  [1, 1], [-1, -1],  # First diagonal
  [1, -1], [-1, 1]   # Second diagonal
].freeze

def valid_position?(row, col, grid)
  row >= 0 && row < grid.length && col >= 0 && col < grid[0].length
end

def chars_at_pattern?(row, col, direction, grid)
  positions = (-1..1).map do |offset|
    [row + (offset * direction[0]), col + (offset * direction[1])]
  end

  return [] unless positions.all? { |pos_row, pos_col| valid_position?(pos_row, pos_col, grid) }

  positions.map { |pos_row, pos_col| grid[pos_row][pos_col] }
end

def pattern_at_position?(row, col, direction, grid)
  chars = chars_at_pattern?(row, col, direction, grid)
  return false if chars.empty?

  [%w[M A S], %w[S A M]].include?(chars)
end

def x_pattern_at?(row, col, grid)
  return false unless grid[row][col] == 'A'

  DIAGONAL_DIRECTIONS.each_slice(2).any? do |dir1, dir2|
    pattern_at_position?(row, col, dir1, grid) &&
      pattern_at_position?(row, col, dir2, grid)
  end
end

def count_x_mas(grid)
  rows = grid.length
  cols = grid[0].length

  (1..rows - 2).sum do |row|
    (1..cols - 2).count do |col|
      x_pattern_at?(row, col, grid)
    end
  end
end

def read_input(input)
  input.split("\n").map(&:chars)
end

def part2(input)
  grid = read_input(input)
  count_x_mas(grid)
end

if __FILE__ == $PROGRAM_NAME
  input = File.read('input.txt')
  puts "Result: #{part2(input)}"
end
